Generalized MOM restart function

marshallward opened the pull request:
This patch merges the internal `save_restart` function with the new `save_MOM6_internal_state` function into a new general MOM restart function.

It also makes an effort to eliminate `MOM_restart` as a driver dependency, narrowing the required MOM API for existing and future drivers.

Also removes the `restart_CSp` argument from `MOM_wave_interface_init`, since it appeared to be used for nothing.

adcroft committed:
Add restart subroutine to MOM.F90

As described in issue #372, I would like to be able to create restart files that
contain information about the particle location. These files will be written at
the same time as other restart files. I cannot add these calls directly to the
driver, because the driver does not have information about the particle location.
We have added save_MOM6_internal_state as a subroutine in MOM.F90, and
we added calls to this subroutine from each of the drivers. We hope this will
allow for more new packages to write restart files in the future.

Co-authored by Spencer Jones <spencerjones@tamu.edu>

marshallward committed:
Generalized MOM restart function

This patch merges the internal `save_restart` function with the new
`save_MOM6_internal_state` function into a new general MOM restart
function.

It also makes an effort to eliminate `MOM_restart` as a driver
dependency, narrowing the required MOM API for existing and future
drivers.

Also removes the `restart_CSp` argument from `MOM_wave_interface_init`,
since it appeared to be used for nothing.

marshallward commented:
I reckon we should get feedback from @adcroft and @hallberg-noaa before adding this to the PR.

adcroft reviewed (commented):
I like this revision. It implements what we discussed the restarts should look like.

marshallward committed:
Create restart directory if absent

MOM simulations typically abort of the restart directory (usually
RESTART) are absent.  This patch adds POSIX support for mkdir() and
creates the directory if it is missing.

marshallward commented:
I added a new feature which creates the `RESTART` directory if it is missing.  The `save_MOM_restart` function feels like the right place to invoke it.

marshallward commented:
@cspencerjones If you like, I can handle this one over at NOAA-GFDL/MOM6, entirely up to you though.

marshallward commented:
These changes will be handled in a separate PR to dev/gfdl.
